docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+### Why is disk accounting necessary?
+
+Present kubernetes clusters become unusable over time due to the local disk becoming full. The kubelets on the node attempt to perform garbage collection of old containers and images, but that doesn’t prevent running pods from using up all the available disk space.
+
+Kubernetes users have no insight into how the disk is being consumed.
+
+Large images and rapid logging, can lead to temporary downtime on the nodes. The node has to free up disk space by deleting images and containers. During this cleanup, existing pods can fail and new pods cannot be started. The node will also transition into an `OutOfDisk` condition, preventing more pods from being scheduled to the node.
+
+Automated eviction of pods that are hogging the local disk is not possible since proper accounting isn’t available.
+
+Since local disk is a non-compressible resource, users need means to restrict usage of local disk by pods and containers. Proper disk accounting is a prerequisite. As of today, a misconfigured low QoS class pod, can end up bringing down the entire cluster by taking up all the available disk space (misconfigured logging for example)
+
+### Goals
+
+1. Account for disk usage on the nodes.
+
+2. Compatibility with the most important docker storage backends - devicemapper, aufs and overlayfs
+
+3. Provide a roadmap for enabling disk as a schedulable resource in the future.
+
+4. Provide a plugin interface for extending support to non-default filesystems and storage drivers.
+
+### Non Goals
+
+1. Compatibility with all storage backends. The matrix is pretty large already and the priority is to get disk accounting to on most widely deployed platforms.
+
+2. Support for filesystems other than ext4.
+
+### Introduction
+
+Disk accounting in Kubernetes cluster running with docker is complex because of the plethora of ways in which disk gets utilized by a container.
+
+Disk can be consumed for,
+
+1. Container images
+
+2. Container’s writable layer
+
+3. Container’s logs - when written to stdout/stderr and default logging backend in docker is used.
+
+4. Local volumes - hostPath and emptyDir (TODO: Do git volumes consume local disk?)
+
+Kubelet exposes disk usage for the entire node and the container’s writable layer for aufs docker storage driver. This information is made available to end users via the heapster monitoring pipeline.
+
+#### Image layers
+
+Image layers are shared between containers (COW) and so accounting for images is complicated.
+
+Image layers will have to be accounted as system overhead.
+
+As of today, it is not possible to perform feasibility checks on images before being pulled on the node.
+
+#### Writable Layer
+
+Docker creates a writable layer for every container on the host. Depending on the storage driver, the location and the underlying filesystem of this layer will change.
```

Comment:
@smarterclayton: Will do. 
@rhvgoyal: Since image layers are shared, image usage needs to be accounted as infrastructure overhead. The writable layer needs to be limited to enforce a total limit on the number of bytes a pod and a container can occupy on local disk.